### PR TITLE
Parameterize binarization function for tradeoff metrics

### LINKED ISSUE #72: Support custom soft label hardening in `TradeoffMetricsRow` calculations
Hacky prototype in #68; now that #69  has landed, we can do it better/for real.

### PR DESCRIPTION
**Pull request by hannahilea**
Closes #72.

- [x] Blocked by #73 merge/tag

### CODE REVIEW

File: src/metrics.jl [hunk 1]
```diff
@@ -882,24 +890,25 @@ function _validate_threshold_class(optimal_threshold_class, classes)
...
                                             elected_hard_labels::AbstractVector, thresholds,
-                                            class_index)
+                                            class_index; harden_fn=harden_by_threshold)
     confusions = [confusion_matrix(2) for _ in 1:length(thresholds)]
     for label_index in 1:length(elected_hard_labels)
         predicted_soft_label = predicted_soft_labels[label_index, class_index]
         elected = (elected_hard_labels[label_index] == class_index) + 1
         for (threshold_index, threshold) in enumerate(thresholds)
-            predicted = (predicted_soft_label >= threshold) + 1
+            predicted = harden_fn(predicted_soft_label, threshold) + 1
```
Comment (ericphanson) [inline src/metrics.jl:911]: entirely unrelated to this change, but this +1 is a little awkward! It makes sense though- going from binary in/out to class 1/2.

File: src/row.jl [hunk 2]
```diff
@@ -169,7 +179,7 @@ function _observation_table_to_inputs(observation_table)
     votes = any(ismissing, df_table.votes) ? missing :
             transpose(reduce(hcat, df_table.votes))
 
-    predicted_soft_labels = transpose(reduce(hcat, df_table.predicted_soft_labels))
+    predicted_soft_labels = _observation_table_to_predicted_soft(df_table.predicted_soft_labels)
```
Comment (ericphanson) [inline src/row.jl:182]: Should it be
```suggestion
    predicted_soft_labels = _predicted_soft_to_matrix(df_table.predicted_soft_labels)
```
?

**Review by ericphanson — COMMENTED** [review]
I wonder if we should call this `binarize` instead? Because I think we use hardening elsewhere to refer to "soft label -> class index" (where class index in 1,2,3,...n_classes) whereas here we want "soft label -> 0,1" (where 0 can't be a class index).

That's something that's currently confusing in different places in lighthouse- when is it a class index vs a binary accept/reject value - maybe calling it binarize can help.

Also, we should add a note in the docstrings as to what it is somewhere. Could do
```julia
const BINARIZE_NOTE = "Supply a function to the keyword argument `binarize` which takes as input `(soft_label, threshold)` and outputs a `Bool` indicating..."
```
and then interpolate it into docstrings to needing to copy-paste it around.

Also, would prefer not include `_fn` in the name; we should just document what is expected (a function doing ...) in the docstring instead.

File: src/metrics.jl (hunk 1 above)
Comment (hannahilea) [reply src/metrics.jl:911]: yeah, agreed 100%

**Comment by hannahilea** [comment]
> I wonder if we should call this binarize instead? ...

Great points/great solutions, updating now!

File: src/metrics.jl (hunk 1 above)
Comment (hannahilea) [reply src/metrics.jl:911]: added a comment

**Review by ericphanson — APPROVED** [review]

**Pull request merged**